Added interpreter for miniscript

sanket1729 opened the pull request:
Added an interpreter for Miniscript which given a satisfying witness to a descriptor, returns the set of public keys, hash preimages, and timelocks that were used. 

I have also added tests to check the "all good" tests. I can add more to check error cases too, but I thought I should get some feedback on the design and code. I will add comments on places where I made some assumptions about the behaviour of the interpreter.

sanket1729 commented on src/descriptor/mod.rs:
```diff
@@ -416,6 +418,31 @@ impl<P: ToPublicKey> Descriptor<P> {
+            Descriptor::Pkh(ref pk) |
+            Descriptor::ShWpkh(ref pk) |
+            Descriptor::Wpkh(ref pk) => {
+                Ok(Primitives { pks: vec![pk.to_public_key()], hashlocks: vec![], timelocks: vec![] })
```
Should the interpreter only check the signatures/hashes inside Miniscript or should it also interpret the entire descriptor and do a check for Pkh(P)? Right now, I am assuming that the descriptor has valid witnesses and only checking the interpretation of Miniscript.

apoelstra replied:
I think it should only check inside the miniscript.

I'd like to have another function which, given a scriptpubkey and witness script, figures out what kind of descriptor it is, but I need to think more about what that might take as input. But that function would be responsible for doing hashchecks.

apoelstra replied:
Yeah, so, it would be cool to have a function
```
pub fn witness_stack(
    script_pubkey: &bitcoin::Script,
    script_sig: &bitcoin::Script,
    witness: Vec<Vec<u8>>,
) -> Result<(Descriptor, Vec<StackElement>), Error>
```
where `StackElement` is an abstract stack element (I suggest a definition for this type in a later comment). This function would (a) check that the witness script is consistent with the scriptpubkey and error out otherwise; (b) figure out what kind of descriptor we're dealing with; (c) translate the `scriptSig` into a set of witness elements for `Bare`/`Pk`/`Pkh`/`Sh` descriptors; (d) drop the witness script from the stack for `Wsh`/`ShWsh` descriptors.

sanket1729 replied:
I am slightly confused by the order in which things are written. I would need to determine (b) in order to solve (a), right?

apoelstra replied:
Yeah, sorry, (a) and (b) should be swapped.

sanket1729 replied:
I have also added the signature validation logic alongside the hash checks for PublicKey based types, `Pkh`/`ShWpkh`/`ShWsh`.

sanket1729 commented on src/descriptor/mod.rs:
```diff
@@ -911,5 +941,374 @@ mod tests {
+                           -> ( Vec<PublicKey>, Vec<Vec<u8> >, secp256k1::Message, Secp256k1<All>) {
+            let secp = secp256k1::Secp256k1::new();
+            let msg = secp256k1::Message::from_slice(
+                &b"Yoda: btc, I trust. HODL I must!"[..]
```
Assumed a standard message for checking signatures

sanket1729 commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,560 @@
+        witness.push(vec![1]);
+    }
+    else{
+        //Even if the script fragment evaluates to false, this method will return the pks, hash
```
If the expression evaluates to false, but still has have valid pukbeys and signatures in it, should those pubkeys be included?
For example, a threshold statement might have valid pubkeys for some subexpressions but does still returns 0.

sanket1729 replied:
The comment in the code does not respect the code below it. I will fix based on the discussion

sanket1729 replied:
This is a general problem for all subexpressions that return 0. So, I am not commenting everywhere.

apoelstra replied:
My feeling is that if the expression evaluates to false, we should ignore any data used to satisfy it. We make an assumption in Miniscript that such data can be 3rd-party-malleated away, so we might as well act as though we did that.

cc @sipa

sanket1729 replied:
Ignored all primitives when the expression evaluated to false

sanket1729 commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,560 @@
+            }
+            AstElem::TimeT(t) => {
+                let res = evaluate_csv(t, age)?;
+                witness.push(build_scriptint(t as i64));
```
I had copy `build_scriptint` from rust bitcoin library as it now a public function and I needed to convert an integer to `Vec<u8>`. This is only for the case where `TimeT` where the return value is not `1`.  

On a similar note, I think all miniscript of `T` form return 1 on success and this is only which is an exception. Is this expression `TimeT` really useful to be added in miniscript?

apoelstra replied:
Yeah, unfortunately there are sane usecases where it saves a byte (for example `and_cat(pk(...), time_t(100))` is the shortest way to put a timelock restriction on `pk(...)` in a T context).

We can do a bit simpler than copying `build_scriptint` though .. since timelocks are always nonnegative we can avoid some of the sign craziness.

apoelstra replied:
I also wonder whether you can just push `true` onto the stack, or even some poison value, since this element should never be popped by any other Miniscript expression.

sanket1729 replied:
I think it is safe to push satisfied instead of int `time` for the current semantics of Miniscript, but could not be the case in the future.  What do you think should be done?

apoelstra replied:
The `T` calling convention is guaranteed to push a truthy value to the stack. It makes no guarantees about what that value is. If this were to change we'd need a new calling convention for that.

sanket1729 replied:
So, you suggest that I push the int instead?

apoelstra replied:
No, I'm suggesting you push `Satisfied`

sanket1729 replied:
Yup, that is the current implementation

apoelstra commented:
I'm not thrilled with the terms "interpret" and "interpreter", since they imply an imperative programming model. But not sure what would be a better way of describing this.

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+fn interpret_and_bool<P>(
+    left: &AstElem<P>,
+    right: &AstElem<P>,
+    witness: &mut Vec<Vec<u8>>,
```
I think that the `witness`, rather than being a `Vec<Vec<u8>>`, should be renamed to `stack`, and be a vector of objects of the type
```
enum StackElement {
    /// Result of a satisfied E/T/W/F
    Satisfied,
    /// Result of a dissatisfied E/W
    Dissatisfied,
    /// Input from the witness stack
    WitnessElement(Vec<u8>),
}
```
rather than using `vec![]` or `vec[1]` the way you would with Script. This way you also eliminate the complexity of using `build_scriptint`.

apoelstra replied:
Given a witness stack, you'd go through a preprocessing stage where you mapped every empty vector to `Satisfied`, every `vec![1]` to `Dissatisfied`, and everything else to `Witness`.

This has the slightly weird effect that missing hash preimages or signatures, which are required to be `vec![]` by Miniscript, will be interpreted as "Dissatisfied", but I think that makes sense.

apoelstra commented on src/descriptor/mod.rs:
```diff
@@ -911,5 +941,374 @@ mod tests {
+            &mut witness,
+            0,
+            &msg,
+            &secp).expect("Interpretation to succeed");
```
Style nit: everywhere you have a `)` on the last line of a multiline thing like this, it would be better to have
```
    ...
    0,
    &msg,
    &secp,
).expect("Interpretation to succeed");
```
Note the trailing comma on `secp` and that the `)` gets its own line.

Let me know if you can cite some style guide saying that your way is correct ... I swear the Rust community changes stuff like this every week.

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+    right: &AstElem<P>,
+    witness: &mut Vec<Vec<u8>>,
+    age: u32,
+    msg: &secp256k1::Message,
```
I'd prefer `msg` be renamed to `sighash` everywhere

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+    witness: &mut Vec<Vec<u8>>,
+    age: u32,
+    msg: &secp256k1::Message,
+    secp: &secp256k1::Secp256k1<All>)
```
Nit: I'd prefer `secp` be the first argument of every function it appears in, rather than the last. You should also use a `Secp256k1<C>` with `C: secp256k1::Verification` rather than `Secp256k1<All>`.

apoelstra replied:
In general I think the order of arguments should be
1. `secp`
2. `stack` (renamed from `witness`)
3. `sighash` (renamed from `msg`)
4. `age`
5. everything else (in this case `left` and `right`)

sanket1729 replied:
Changed

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+    let leftres = witness.pop().expect("Witness stack has at least one element");
+    let rightres = witness.pop().expect("Witness stack has at least one element");
+    let orres = leftres == vec![1] || rightres == vec![1];
+    if orres{
```
style nit: space before `{`

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+    if orres{
+        witness.push(vec![1]);
+    }
+    else{
```
style nit: `} else {` should be one line

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,561 @@
+    age: u32,
+    msg: &secp256k1::Message,
+    secp: &secp256k1::Secp256k1<All>)
+    -> Result<Primitives, Error >
```
Style nit: `)` should be at the beginning of the line, like `) -> Result<Primitives, Error>`, and there should be a comma after `Secp256k1<All>` above

sanket1729 commented:
Addressed all the above comments and coding style nits. The PR is again ready for review. 

@apoelstra  This will require a rebase after merging #8  and #11, maybe we get those in first.

sanket1729 commented:
Interpreter can also be changed to `Evaluator` if that sounds better

apoelstra commented:
"evaluate" is even worse than "interpret" :) the point is that we aren't executing code, we're determining which data was used to satisfy the Miniscript conditions

apoelstra commented on src/descriptor/mod.rs:
```diff
@@ -25,18 +25,22 @@
+use secp256k1::VerifyOnly;
 
+
+pub mod create_descriptor;
```
Rather than `pub mod create_descriptor` can you do
```
mod create_descriptor;
pub use create_descriptor::witness_stack;
```
so that `create_descriptor` is not exposed in the API as a separate module?

apoelstra commented on src/descriptor/create_descriptor.rs:
```diff
@@ -0,0 +1,446 @@
+//! 6) ShWpkh: Sh: Checks redeem_script hash, translates remaining elements from script_sig to
+//! witness_stack<StackElement> and validates Wpkh sig, pubkey.
+//! 7) ShWsh: Checks witness script hash, pops witness script and converts it to miniscript.
+//! translates the remaining witness to witness_stack<StackElement>
```
1. I think this whole block comment should actually go on `pub fn witness_stack` rather than on the module.
2. I don't think we should be verifying signatures in this function; that seems like a responsibility of the interpreter (and it seems like you're already doing it there)

sanket1729 replied:
The signatures are being checked for non-miniscript `Pkh/WPkh/ShWpkh` descriptors. 

For `Pkh/WPkh/ShWpkh` types, I should only check the pubkey hash matches and not the signatures. And the interpreter should then check only the signatures?

apoelstra replied:
I think so, yeah. Reason is that I want to be able to abuse `witness_stack` by giving it only the witnessScript/redeemScript and no actual satisfying data, and have it give me the `Descriptor` and an empty `Vec<StackElement>` back.

This gives me the ability to get a descriptor from a scriptpubkey/scriptsig/witness script.

sanket1729 replied:
So, `Pkh/WPkh/ShWpkh` should return empty `Vec<StackElement>` or vec![sig pk] or vec![sig] since is in descriptor.

sanket1729 replied:
Sorry, understood what you meant finally with the above comment. 

For `Pkh` if only pk is supplied as `witness` without a sig. It will return empty. If it provided with a `sig`, that will return `[StackElement::Witness(sig)]`. A similar logic for `Wpkh/Shwpkh`

apoelstra replied:
Yep exactly :)

apoelstra commented on src/descriptor/create_descriptor.rs:
```diff
@@ -0,0 +1,446 @@
+    match ins {
+        Instruction::PushBytes(v) => StackElement::from(v.to_vec()),
+        Instruction::Op(opcodes::all::OP_PUSHNUM_1) => StackElement::Satisfied,
+        _ => unreachable!(),
```
Why don't we translate empty pushes to `Dissatisfied`?

Also we can't panic on non-push data because we're starting from a scriptsig/witness stack which is user-controlled data. We should return an error instead.

sanket1729 replied:
`Dissatisfied` is actually translated as PUSHBYTES_0 which converted from `StackElement::from(v.to_vec()` . Will change the unreachable to error.

apoelstra replied:
Ah, nice. Can you add a comment above `PushBytes(v)` line?

sanket1729 commented:
Addressed the newer comments.

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,606 @@
+    Dissatisfied,
+    /// Input from the witness stack
+    Witness(Vec<u8>),
+}
```
Can you try
```

#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum StackElement<'a> {
    /// Result of a satisfied E/T/W/F, or input to `IF`/`NOTIF`
    Satisfied,
    /// Result of a dissatisfied E/W, or input to `IF`/`NOTIF`
    Dissatisfied,
    /// Other data from the witness stack
    Push(&'a [u8]),
}
```
You'll run into a bunch of weird lifetime issues doing the `'a` thing - if you have too much trouble let me know and I'll add a commit which does this. But the upside is that you'll avoid a bunch of extra allocations.

But at the very least can you rename `Witness` to `Push` (my, bad!). I didn't realize when I made my original suggestion that `Satisfied` and `Dissatisfied` might also be input from the witness stack, so naming one variant `Witness` is a bit confusing.

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,606 @@
+
+impl StackElement{
+
+    pub fn is_witness_or_err(self) -> Result<Vec<u8>, Error>
```
I think you should implement this on `Option<StackElement>`, have it take `&self` rather than `self`, and return a `&[u8]` rather than `Vec<u8>`.

apoelstra replied:
and rename the method to just `witness`. The `_or_err` is implied by the type.

apoelstra replied:
Similarly you should have a method `preimage(hash: sha256d::Hash)` which returns a preimage if it satisfies the length/hash check

apoelstra commented on src/descriptor/mod.rs:
```diff
@@ -416,6 +419,37 @@ impl<P: ToPublicKey> Descriptor<P> {
+            Descriptor::Pkh(ref pk) |
+            Descriptor::ShWpkh(ref pk) |
+            Descriptor::Wpkh(ref pk) => {
+                let sig = stack.pop().expect("Signature").is_witness_or_err()?;
```
This `expect` is invalid because there might not be anything on the stack when the user calls this function. See my below comment about `is_witness_or_err` - I think you should give the `Option<_>` from `pop` directly to that method.

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,606 @@
+///Primitives for Miniscript: a vector of public keys(sigs), hashlocks and timelocks
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Primitives{
+    pub pks: Vec<bitcoin::PublicKey>,
```
I think we should parameterize this structure over the public key type. Also it'd be cool to just return references into the descriptor rather than copying everything, like

```
pub struct Primitives<'desc, 'stack, P> {
    pub pks: Vec<&'desc P>,
    pub hashlocks: Vec<(&'stack [u8], sha256d::Hash)>,
    pub timelocks: Vec<u32>,
}
```
and you should also have doccomments (even trivial/redundant ones) on all the fields

apoelstra replied:
Also, I think `pks` should be a `HashSet` (which puts a `std::hash::Hash` requirement on `P`, which kinda sucks, but should be fine in practice), `hashlocks` should be a `HashMap<&'desc sha256d::Hash, &'a stack[u8]>`, and `timelocks` should also be a `HashSet`

apoelstra replied:
There are a few advantages of using references like this

1. Less allocation/copying
2. It gives the user a way to determine which part of the descriptor a public key came from, in case there are multiple instances of the same key (by casting to `*const _` and comparing the pointers)
3. It improves documentation because it's clear where each piece of data comes from

apoelstra replied:
Actually maybe we want to change the API entirely to

```
/// A satisfied Miniscript condition (e.g. signature requirement)
enum SatisfiedConstraint<'desc, 'stack, P> {
    PublicKey {
        key: &'desc P,
        sig: secp256k1::Signature,
    },
    HashLock {
        hash: &'desc sha256d::Hash,
        preimage: &'stack [u8],
    },
    RelativeTimeLock {
        time: &'desc u32,
    },
}
```
then
```
/// An iterator over all the constraints satisfied by a given
/// descriptor/scriptSig/witness stack tuple
pub struct SatisfiedConstraints {
    // ...
}

impl<'desc, 'stack, P> Iterator for SatisfiedConstraints {
    type Item = SatisfiedConstraint<'desc, 'stack, P>;

    fn next(&mut self) -> Option<SatisfiedConstraint<'desc, 'stack, P>> {
        // ...
    }
}
```
Does this make sense?

apoelstra replied:
Then `Descriptor::interpret` could be replaced by a function `SatisfiedConstraints::from_descriptor<'desc, 'stack>(descriptor: &'desc Descriptor, stack: Vec<StackElement<'stack>>)` which creates this iterator, (and all of the code could exist in its own module, perhaps reexported as part of the public `descriptor` module).

apoelstra commented on src/miniscript/evaluate.rs:
```diff
@@ -0,0 +1,606 @@
+/// This functions is equivalent to VERIFY. This pops the top element and checks if it is non-zero.
+/// Gives an error in all other cases
+fn evaluate_verify(stack: &mut Vec<StackElement>) -> Result<(), Error>{
+    let top = stack.pop().expect("Non-zero element for verify");
```
These `expect`s all make me nervous. Suggest changing the whole function body to
```
stack.pop() == StackElement::Satisfied
```
(This is also a logic change, from "anything but `Dissatisfied`" to "must be `Satisfied`", which I think is more correct ... in particular there are some non-`Dissatisfied` pushes that the Script interpreter will evaluate as false, and in general I'd prefer to be as strict as we can possibly get away with.)

apoelstra replied:
I also think (though maybe you disagree) that it'd look a bit nicer to `impl` all these methods on `Vec<StackElement>`, like
```
impl Vec<StackElement> {
    fn verify(&mut self) -> Result<(), Error> {
        if stack.pop() == Some(StackElement::Satisfied) {
            Ok(())
        } else {
            Err(Error::CouldNotEvaluate)
        }
    }
}
```

apoelstra commented on src/descriptor/create_descriptor.rs:
```diff
@@ -0,0 +1,410 @@
+            //ensures that scriptsig after popping redeem script contains 0 elements.
+            //Therefore while calling verify_wpkh, an argument of Script::new() is passed instead
+            //of script_sig. The redeem_script becomes the script_pubkey
+            if stack.len() != 0{
```
nit: it's a bit more idiomatic to use `stack.is_empty()` rather than comparing `len` to 0. (And also for some non-`Vec` collections it's much more efficient).

This is something `clippy` will warn you about BTW ... if it's not too noisy (as I mentioned, it hasn't been compiling for me lately so I haven't used it) it's definitely worth running.

apoelstra commented on src/descriptor/create_descriptor.rs:
```diff
@@ -0,0 +1,410 @@
+/// Converts the other script elements into Vec<StackElement>
+fn parse_scriptsig_top(script_sig: &bitcoin::Script) -> Result<(Vec<u8>, Vec<StackElement>), Error>
+{
+    let instructions : Vec<Instruction> = script_sig.iter(false).collect();
```
I think we should use `iter(true)` everywhere, as part of our "be as strict as possible" ethos.

sanket1729 commented:
Opening a new PR for this based on new version of miniscript
